File: cil-tools/task_struct.c
/* This is an example import file, showing how to import the
 * Linux task_struct structure from header files.
 *
 * Use: bitstring-import-c task_struct.c > task_struct.ml
 *
 * Tip: Add the --debug flag to that command line to see what's going on.
 * Also use bitstring-objinfo to examine the bmpp file.
 */

/* Any defines, etc. necessary to get the include to work. */
#define __KERNEL__
#define CONFIG_HZ 100
#define CONFIG_PAGE_OFFSETUL 0xc0000000 /* XXX? */
#define THREAD_SIZE 4096 /* XXX? */

/* Include files necessary to get the structure(s) and constant(s) we're
 * interested in.
 *
 * Note in this case glibc strips the useful structures out of the
 * so-called "kernel headers" that it supplies, so instead I have
 * a local copy of the real headers from a version of Linux.
 */
#include "linux/sched.h"

/* This tells the importer program what structures and constants to import. */
typedef struct task_struct BITSTRING_IMPORT(task_struct);

File: cil-tools/task_struct.ml
(* This is an example program using an imported C structure.
 *
 * The Linux process table entry structure is imported from
 * <linux/sched.h> via "task_struct.c" by the bitmatch-import-c
 * program, and saved as "task_struct.bmpp".
 *
 * Then we can load "task_struct.bmpp" here.
 *)

open Printf

open bitmatch "task_struct.bmpp"

(*
let () =
  let bits = Bitmatch.bitstring_of_file "examples/ext3_sb" in
*)
